src/svm/impls/binary/binary.c

```diff
@@ -1411,141 +1405,95 @@ PetscErrorCode SVMGetModelScore_Binary(SVM svm,ModelScore score_type,PetscReal *
+
+    n_1 = (PetscAbs(end_1 - start_1) + 1) / PetscAbs(step_1);
+    n_2 = (PetscAbs(end_2 - start_2) + 1) / PetscAbs(step_2);
+    n = 2 * n_1 * n_2;
+    TRY( PetscMalloc1(n,&grid) );
+
+    p = -1;
+    for (i = 0; i < n_1; ++i) {
+      grid[++p] = PetscPowReal(base_1,start_1 + i * step_1);
+      tmp = grid[p];
+      grid[++p] = PetscPowReal(base_2,start_2);
+      for (j = 1; j < n_2; ++j) {
+        grid[++p] = tmp;
+        grid[++p] = PetscPowReal(base_2,start_2 + j * step_2);
       }
-      Cp *= logCp_base;
     }
   }
 
-  *n = n_inner;
-  *c_arr = c_arr_inner;
+  *n_out = n;
+  *grid_out = grid;
   PetscFunctionReturn(0);
 }
 
 #undef __FUNCT__
 #define __FUNCT__ "SVMGridSearch_Binary"
 PetscErrorCode SVMGridSearch_Binary(SVM svm)
 {
-  MPI_Comm   comm;
-
-  PetscReal  *c_arr,*score;
-  PetscReal  score_best;
-  PetscInt   m,n,i,p;
-
-  PetscBool  info_set;
-  const char *prefix;
-
-  const ModelScore *hyperopt_score_types;
-  PetscInt   nscores;
+  PetscReal *grid;
+  PetscReal *scores,score_best;
+  PetscInt  i,n,m,p,s;
 
   PetscFunctionBegin;
-  TRY( SVMInitGridSearch_Binary_Private(svm,&n,&c_arr) );
   TRY( SVMGetPenaltyType(svm,&m) );
 
-  TRY( PetscMalloc1((n / m),&score) );
-  TRY( PetscMemzero(score,(n / m) * sizeof(PetscReal)) );
-
-  TRY( SVMCrossValidation(svm,c_arr,n,score) );
-
-  /* Select penalty */
-  n /= m;
-  score_best = score[0];
-  p = 0;
-  for (i = 1; i < n; ++i) {
-    if (score[i] > score_best) {
+  /* Initialize grid */
+  TRY( SVMInitGridSearch_Binary_Private(svm,&n,&grid) );
+  /* Perform cross-validation */
+  s = n / m;
+  TRY( PetscMalloc1(s,&scores) );
+  TRY( PetscMemzero(scores,s * sizeof(PetscReal)) );
+  TRY( SVMCrossValidation(svm,grid,n,scores) );
+
+  /* Find best score */
+  score_best = -1.;
+  for (i = 0; i < s; ++i) {
+    if (scores[i] > score_best) {
       p = i;
-      score_best = score[i];
+      score_best = scores[i];
     }
   }
-  TRY( SVMSetPenalty(svm,m,&c_arr[p * m]) );
-
-  TRY( SVMGetOptionsPrefix(svm,&prefix) );
-  TRY( PetscOptionsHasName(NULL,prefix,"-svm_info",&info_set) );
-
-  if (info_set) {
-    TRY( PetscObjectGetComm((PetscObject) svm,&comm) );
-    TRY( SVMGetHyperOptScoreTypes(svm,&hyperopt_score_types) );
-    TRY( SVMGetHyperOptNScoreTypes(svm,&nscores) );
-    TRY( PetscPrintf(comm,"SVM (grid-search): selected ") );
-    if (m == 1) {
-      TRY( PetscPrintf(comm,"C_best=%f, ",c_arr[p]) );
-    } else {
-      TRY( PetscPrintf(comm,"C+_best=%f, ",c_arr[p * m]) );
-      TRY( PetscPrintf(comm,"C-_best=%f, ",c_arr[p * m + 1]) );
-    }
-    TRY( PetscPrintf(comm,"acc_score=%f (",score_best) );
-    for (i = 0; i < nscores; ++i) {
-      TRY( PetscPrintf(comm,"%s",ModelScores[hyperopt_score_types[i]]) );
-      if (i < nscores - 1) {
-        TRY( PetscPrintf(comm,",") );
-      } else {
-        TRY( PetscPrintf(comm,")\n") );
-      }
-    }
+
+  if (m == 1) {
+    PetscInfo2(svm,"selected best C=%.4f (score=%f)\n",grid[p],score_best);
```

Comment:
The issue also appears in other parts of code. I think it is better to do it in another PR related to fixing code-style.

Reply:
I think we should not introduce a potential bug (not casting will be undefined behaviour when PetscReal is not double/float?) just because it is somewhere else in the code.

I added note in #3 to fix this in other parts of the code.